pkg/webhook/server.go

```diff
@@ -53,256 +52,145 @@ type ServerOptions struct {
+	setFields inject.Func
 
-	// Deprecated: Writer will not be used anywhere.
-	Writer io.Writer
-
-	// Service is k8s service fronting the webhook server pod(s).
-	// This field is optional. But one and only one of Service and Host need to be set.
-	// This maps to field .webhooks.getClientConfig.service
-	// https://github.com/kubernetes/api/blob/183f3326a9353bd6d41430fc80f96259331d029c/admissionregistration/v1beta1/types.go#L260
-	Service *Service
-	// Host is the host name of .webhooks.clientConfig.url
-	// https://github.com/kubernetes/api/blob/183f3326a9353bd6d41430fc80f96259331d029c/admissionregistration/v1beta1/types.go#L250
-	// This field is optional. But one and only one of Service and Host need to be set.
-	// If neither Service nor Host is unspecified, Host will be defaulted to "localhost".
-	Host *string
-
-	// certProvisioner is constructed using certGenerator and certWriter
-	certProvisioner *cert.Provisioner // nolint: structcheck
-
-	// err will be non-nil if there is an error occur during initialization.
-	err error // nolint: structcheck
+	// defaultingOnce ensures that the default fields are only ever set once.
+	defaultingOnce sync.Once
 }
 
-// Service contains information for creating a service
-type Service struct {
-	// Name of the service
-	Name string
-	// Namespace of the service
-	Namespace string
-	// Selectors is the selector of the service.
-	// This must select the pods that runs this webhook server.
-	Selectors map[string]string
+// setDefaults does defaulting for the Server.
+func (s *Server) setDefaults() {
+	s.webhooks = map[string]http.Handler{}
+	s.webhookMux = http.NewServeMux()
+	if s.Port <= 0 {
+		s.Port = 443
+	}
+	if len(s.CertDir) == 0 {
+		s.CertDir = path.Join("/tmp", "k8s-webhook-server", "serving-certs")
+	}
 }
 
-// Server is an admission webhook server that can serve traffic and
-// generates related k8s resources for deploying.
-type Server struct {
-	// Name is the name of server
-	Name string
-
-	// ServerOptions contains options for configuring the admission server.
-	ServerOptions
-
-	sMux *http.ServeMux
-	// registry maps a path to a http.Handler.
-	registry map[string]Webhook
-
-	// mutatingWebhookConfiguration and validatingWebhookConfiguration are populated during server bootstrapping.
-	// They can be nil, if there is no webhook registered under it.
-	webhookConfigurations []runtime.Object
-
-	// manager is the manager that this webhook server will be registered.
-	manager manager.Manager
-
-	// httpServer is the actual server that serves the traffic.
-	httpServer *http.Server
-
-	once sync.Once
+// For registers defaulting and validation webhooks for the provided types that implement either
+// webhooktuil.Validator or webhookutil.Defaulter.
+func (s *Server) For(objects ...runtime.Object) error {
+	// Create a webhook for each type
+	for _, object := range objects {
+		wh := NewResourceWebhook(object)
+		if wh == nil {
+			continue
+		}
+		s.Register("path", wh)
```

Comment:
We can generate the path using object's GVK and name. And, the default path generation should probably be shared with controller-tools so that generator is in sync with this code ?